ipaserver/install/dogtaginstance.py

```diff
@@ -275,6 +279,64 @@ def uninstall(self):
             logger.critical("failed to uninstall %s instance %s",
                             self.subsystem, e)
 
+    def secure_ajp_connector(self):
+        """ Update AJP connector to use a password protection  """
+        server_xml = open(paths.PKI_TOMCAT_SERVER_XML, 'r').read()
+
+        # AJP connector is set on port 8009. Use non-greedy search to find it
+        connector = re.search('<Connector port="8009".+?/>', server_xml)
+
+        # no AJP connector means no need to update anything
+        if connector is None:
+            return
+
+        # Detect tomcat version and choose the right option name
+        # pre-9.0.31.0 uses 'requiredSecret'
+        # 9.0.31.0 or later uses 'secret'
+        version = ('pre-9.0.31.0', ' requiredSecret="')
+        try:
+            result = ipautil.run([paths.BIN_TOMCAT, "--version"],
+                                 capture_output=True)
+            sn = re.search(
+                r'Server number:\s+([0-9]+\.[0-9]+\.[0-9]+\.[0-9]+)$',
+                result.output, flags=re.MULTILINE)
+            if sn is None:
+                logger.info("Unsupported tomcat version, default to %s",
+                            version[0])
+            else:
+                v = tasks.parse_ipa_version(sn.group(1))
+                if v >= tasks.parse_ipa_version('9.0.31.0'):
+                    version = (sn.group(1), ' secret="')
+        except ipautil.CalledProcessError as e:
+            logger.info(
+                "failed to discover tomcat version, default to %s, error: %s",
+                version[0], str(e))
+
+        secret = re.search('(secret|requiredSecret)="(.+?)"',
+                           connector.group(0))
+        # no secret is in place, add a new one
+        if secret:
+            # secret is already in place
+            # Perhaps, we need to synchronize it with Apache configuration
+            self.ajp_secret = secret.group(2)
+        else:
+            self.ajp_secret = ipautil.ipa_generate_password()
+
+        # Write new server.xml if we have to add the secret or
+        # if we have to update the option name
+        if (secret and secret.group(1) not in version[1]) or secret is None:
```

Comment:
Won't this allow both secret="something" and requiredSecret="something"?

Reply:
The current proposed code seems to work just fine, I'm now debugging KRA instance failures.